master/internal/experiment.go

```diff
@@ -490,6 +429,72 @@ func (e *experiment) Receive(ctx *actor.Context) error {
 	return nil
 }
 
+// snapshot snapshots the experiment and saves it in the database.
+func (e *experiment) snapshot(t trialSnapshot) error {
+	if ss, err := e.Snapshot(); err != nil {
+		return errors.Wrap(err, "failed to snapshot experiment")
+	} else if err := e.db.SaveSnapshot(e.ID, t.trialID, t.requestID, ss, t.snapshot); err != nil {
+		return errors.Wrap(err, "failed to persist experiment snapshot")
+	}
+	return nil
+}
+
+// restore restores this experiment from a snapshot in the database if it exists.
+func (e *experiment) restore() error {
+	switch b, err := e.db.ExperimentSnapshot(e.ID); {
+	case errors.Cause(err) == db.ErrNotFound:
+		log.WithField("experiment-id", e.ID).Info("no snapshot found")
+	case err != nil:
+		return errors.Wrap(err, "failed to retrieve experiment snapshot")
+	default:
+		if err := e.Restore(b); err != nil {
+			return errors.Wrap(err, "failed to restore experiment snapshot")
+		}
+	}
+	return nil
+}
+
+func (e *experiment) restoreTrial(
```

Comment:
seems like a lot of this stuff should go in trial.go and happen more-or-less automatically in newTrial.

Reply:
i think we talked about this offline: it was this way (and restoreTrial is on the experiment and not in newTrial or restore on trial) because it is experiment that is doing the restoring. it's usually its context to log, send messages, etc. since it needs to send some messages it restore and it feels weird to give its context to a trial on start, im not sure if it can go in trial. but one i clean everything else up i'll push the code around until it looks a bit better (which may include moving this)